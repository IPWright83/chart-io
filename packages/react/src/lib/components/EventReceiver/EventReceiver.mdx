import { Meta, Story, Canvas } from "@storybook/blocks";

<Meta title="Components/EventReceiver" />

# EventReceiver Component

The `<EventReceiver>` component is an internal component used for capturing mouse events on SVG based charts.

Some plots do not provide a suitable element to capture events on (such as a Line chart) so instead they are captured on this component and events are dispatched to the internal redux store - giving plots a chance to respond to these changes.

<Canvas>
    <Story id="components-eventreceiver--default" />
</Canvas>
