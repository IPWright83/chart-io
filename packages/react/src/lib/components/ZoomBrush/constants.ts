export const DEFAULT_BRUSH_MARGIN = { left: 0, top: 10, right: 0, bottom: 10 };
