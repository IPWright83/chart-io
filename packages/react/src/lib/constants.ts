export const PROGRESSIVE_RENDER_THRESHOLD = 5000;
export const MOUSE_MOVE_THROTTLE = 50;
export const PACKAGE_NAME = "@chart-it/react";
export const STORYBOOK_URL = "https://ipwright83.github.io/chart-it/";
