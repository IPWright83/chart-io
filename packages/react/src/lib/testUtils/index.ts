export * from "./actionsIncludes";
export * from "./FakeMouseEvent";
export * from "./createMockStore";
export * from "./getBuffer";
export * from "./wait";
export * from "./renderChart";
export * from "./testEvents";
