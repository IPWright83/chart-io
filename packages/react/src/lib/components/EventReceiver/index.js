import { withSVG } from "../../hoc";
import { EventReceiver as EventReceiverBase } from "./EventReceiver";

const EventReceiver = withSVG(EventReceiverBase, "event-receiver");

export { EventReceiver };
