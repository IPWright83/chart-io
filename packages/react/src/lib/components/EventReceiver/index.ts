import { withSVG } from "../../hoc";
import { EventReceiver as EventReceiverBase, IEventReceiverBaseProps } from "./EventReceiver";

export interface IEventReceiverProps extends Omit<IEventReceiverBaseProps, "layer"> {}

export const EventReceiver = withSVG<IEventReceiverProps>(EventReceiverBase, "event-receiver");
