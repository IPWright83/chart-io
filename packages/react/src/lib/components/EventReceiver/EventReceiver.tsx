import * as d3 from "@d3-chart/d3";
import { throttle } from "lodash";
import React, { useEffect } from "react";
import { useStore, useSelector } from "react-redux";

import { chartSelectors, eventActions, IState } from "../../store";
import { MOUSE_MOVE_THROTTLE } from "../../constants";

export interface IEventReceiverBaseProps {
    /**
     * The layer to be rendered upon. Typically this is an `<svg:g>` or a fake HTMLElement when using canvas.
     * @default undefined
     */
    layer: React.MutableRefObject<Element>;
}

/**
 * Creates a background layer which is used to capture mouse events
 * @return The EventReceiver component
 */
export function EventReceiver({ layer }: IEventReceiverBaseProps) {
    const store = useStore();
    const width = useSelector((s: IState) => chartSelectors.dimensions.width(s));
    const height = useSelector((s: IState) => chartSelectors.dimensions.height(s));
    const margin = useSelector((s: IState) => chartSelectors.dimensions.margin(s));

    // Runs whenever the dimensions change
    useEffect(() => {
        if (!layer.current || !width || !height) {
            return;
        }

        const mouseMove = throttle(
            (e) => {
                store.dispatch(eventActions.mouseMove(e));
            },
            MOUSE_MOVE_THROTTLE,
            { leading: true }
        );

        // prettier-ignore
        d3.select(layer.current)
            .select("rect")
            .attr("width", width - margin.left - margin.right)
            .attr("height", height - margin.top - margin.bottom)
            .on("mouseout", (e) => { store.dispatch(eventActions.mouseExit(e)); })
            .on("mouseover", (e) => { store.dispatch(eventActions.mouseEnter(e)); })
            .on("mousemove", mouseMove);

        // Wire up events
    }, [store.dispatch, layer, width, height, margin]);

    const transform = `translate(${margin.left || 0}, ${margin.top || 0})`;
    const style = { fill: "none", pointerEvents: "all" as const };

    return <rect className="d3-chart event-receiver" transform={transform} style={style} />;
}
