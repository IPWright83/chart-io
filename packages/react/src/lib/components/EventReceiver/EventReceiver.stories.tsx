import React from "react";
import { Provider } from "react-redux";

import { themes } from "../../themes";
import { EventReceiver } from ".";

import { createMockStorybookStore } from "../../testUtils";

export default {
    title: "Components/EventReceiver",
    component: EventReceiver,
    parameters: {
        chromatic: { delay: 300 },
    },
    argTypes: {
        mouseEvent: { action: "mouse" },
    },
};

const EventReceiverTemplate = (args) => {
    const store = createMockStorybookStore({
        chart: {
            dimensions: {
                width: 200,
                height: 200,
            },
            theme: themes.light,
        },
    });

    store.dispatch = console.log; //args.mouseEvent;

    return (
        <div>
            <p>Move your cursor over the blue box and check the console to see events</p>
            <div style={{ backgroundColor: "steelblue" }}>
                <Provider store={store}>
                    <svg>
                        <EventReceiver />
                    </svg>
                </Provider>
            </div>
        </div>
    );
};

export const Default = EventReceiverTemplate.bind({});
Default.storyName = "EventReceiver";
