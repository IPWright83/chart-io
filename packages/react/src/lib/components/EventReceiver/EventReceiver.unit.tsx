import { fireEvent, render } from "@testing-library/react";
import { Provider } from "react-redux";
import React from "react";

import { createMockStore, FakeMouseEvent, wait } from "../../testUtils";
import { MOUSE_MOVE_THROTTLE } from "../../constants";

import { EventReceiver } from ".";
import { EventReceiver as EventReceiverBase } from "./EventReceiver";

describe("EventReceiver", () => {
    const store = createMockStore({
        chart: {
            dimensions: {
                width: 800,
                height: 400,
                margin: { left: 10, right: 20, top: 30, bottom: 40 },
            },
        },
    });

    it("should render correctly", async () => {
        const { asFragment } = render(
            <Provider store={store}>
                <svg>
                    <EventReceiver />
                </svg>
            </Provider>
        );

        expect(asFragment()).toMatchSnapshot();
    });

    it("should trigger throttled mousemove event", async () => {
        const { container } = render(
            <Provider store={store}>
                <svg>
                    <EventReceiver />
                </svg>
            </Provider>
        );

        jest.spyOn(store, "dispatch");

        const element = container.querySelector("rect");
        fireEvent(element, new FakeMouseEvent("mouseover", { pageX: 150, pageY: 150 }));
        fireEvent(element, new FakeMouseEvent("mousemove", { pageX: 150, pageY: 150 }));
        fireEvent(element, new FakeMouseEvent("mouseout", { pageX: 900, pageY: 900 }));

        await wait(2 * MOUSE_MOVE_THROTTLE);

        const dispatchCalls = (store.dispatch as jest.Mock).mock.calls.map((c) => c[0].type);
        expect(dispatchCalls).toEqual(["EVENT.MOUSE_ENTER", "EVENT.MOUSE_MOVE", "EVENT.MOUSE_EXIT"]);
    });

    it("should skip if there is no layer avaliable", () => {
        const layer = { current: null };

        render(
            <Provider store={store}>
                <svg>
                    <EventReceiverBase layer={layer} />
                </svg>
            </Provider>
        );

        // Should be empty
        expect(layer).toMatchSnapshot();
    });
});
