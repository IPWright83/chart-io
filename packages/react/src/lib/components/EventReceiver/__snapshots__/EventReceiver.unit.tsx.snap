// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EventReceiver should render correctly 1`] = `
<DocumentFragment>
  <svg>
    <g
      class="g-event-receiver"
    >
      <rect
        class="chart-it event-receiver"
        height="330"
        style="fill: none; pointer-events: all;"
        transform="translate(10, 30)"
        width="770"
      />
    </g>
  </svg>
</DocumentFragment>
`;

exports[`EventReceiver should skip if there is no layer avaliable 1`] = `
{
  "current": null,
}
`;
